File: 2023/UICPC Regional/Round #15/solutions/jealousyoungsters/per.py
(n, m) = map(int, raw_input().split())
(b, r) = map(int, raw_input().split())
totplay = [[0]*m for _ in range(n)]
firstplay = [[b+1]*m for _ in range(n)]
curtoy = [(-1, 0)]*n
curkid = [(-1, 0)]*m
for (s, k, t) in [map(int, raw_input().split()) for _ in range(r)]:
    k -= 1
    t -= 1
    if t != -1: firstplay[k][t] = min(firstplay[k][t], s)
    (pt, ps) = curtoy[k]
    if pt == t: continue
    if pt != -1:
        totplay[k][pt] += s - ps
        curkid[pt] = (-1, 0)
    if t != -1:
        (pk, ps) = curkid[t]
        if pk != -1:
            totplay[pk][t] += s - ps
            curtoy[pk] = (-1, 0)
        curkid[t] = (k, s)
    curtoy[k] = (t, s)
for i in range(n):
    if curtoy[i][0] != -1: totplay[i][curtoy[i][0]] += b - curtoy[i][1]

if m < n:
    print 'impossible'
else:
    wish = [iter(sorted(range(m), key = lambda t: firstplay[k][t])) for k in range(n)]
    kid = [-1]*m
    toy = [0]*n
    for k in range(n):
        while k >= 0:
            t = next(wish[k])
            if kid[t] < 0 or totplay[k][t] < totplay[kid[t]][t]:
                toy[k] = t+1
                (k, kid[t]) = (kid[t], k)
    print ' '.join(map(str, toy))

File: 2023/UICPC Regional/Round #15/solutions/jealousyoungsters/per.cc
#include <bits/stdc++.h>
using namespace std;

int main(void) {
    int n, m, b, r;
    scanf("%d%d%d%d", &n, &m, &b, &r);
    if (m < n) {
        printf("impossible\n");
        return 0;
    }
    vector<vector<int>> totplay(n, vector<int>(m, 0));
    vector<vector<int>> firstplay(n, vector<int>(m, b+1));
    vector<pair<int, int>> curtoy(n, make_pair(-1, 0));
    for (int i = 0; i < r; ++i) {
        int s, k, t;
        scanf("%d%d%d", &s, &k, &t);
        --k; --t;
        if (t != -1) firstplay[k][t] = min(firstplay[k][t], s);
        auto p = curtoy[k];
        if (p.first != -1) totplay[k][p.first] += s - p.second;
        curtoy[k] = make_pair(t, s);
    }
    for (int i = 0; i < n; ++i)
        if (curtoy[i].first != -1)
            totplay[i][curtoy[i].first] += b - curtoy[i].second;
    vector<vector<int>> wish(n, vector<int>(m, 0));
    for (int i = 0; i < n; ++i) {
        iota(wish[i].begin(), wish[i].end(), 0);
        sort(wish[i].begin(), wish[i].end(), [&](int a, int b) {
                return firstplay[i][a] > firstplay[i][b];
            });
        assert(wish[i].size() == m);
    }
    vector<int> kid(m, -1), toy(n, 0);
    for (int _k = 0; _k < n; ++_k) {
        int k = _k;
        while (k >= 0) {
            assert(!wish[k].empty());
            int t = wish[k].back(); wish[k].pop_back();
            if (kid[t] < 0 || totplay[k][t] < totplay[kid[t]][t]) {
                toy[k] = t+1;
                swap(k, kid[t]);
            }
        }
    }
    for (int i = 0; i < n; ++i)
        printf("%d\n", toy[i]);
}
